### PR TITLE
feat: Logging `GoalState` objects

### PR DESCRIPTION
**Pull request by scottcanoe**
**Summary**

This PR consists of the following changes:
 1. The addition of an encoder for `GoalState` objects to allow them to be serialized.
 2. The addition of an `info` attribute for `GoalState` objects which is meant to hold only log-bound information.
 3. The insertion of `GoalState` objects into the buffer.
 4. A small test unit test of the goal state encoder has been added, and one line in a testing utility has been modified to prevent accidentally indexing into an empty list.

**Motivation**
The Demonstrating Monty's Capabilities paper requires currently unlogged goal state information. `tbp.monty` needed an update to accommodate this. However, the ability to log goal states--including attributes used to generate them--will also be a very useful feature for our upcoming work on new motor policies.

**Discussion**
The changes introduced here are kept to a minimum, but I wanted to highlight an obvious follow-up regarding vanilla `State` objects.

  1. `State` objects are not currently stored in a buffer, I'm assuming due to the inability to serialize them later. Instead, all of their attributes are appended to lists held by the buffer. A desire for a cleaner approach has been hinted at in a TODO in `buffer.py` line 113:
      ```python
          """Add an observation to the buffer. Must be features at locations.
  
          TODO S: Store state objects instead of list of data?
          """
      ```
  2. Log-only attributes are already being attached to `State` objects, but they are not obviously marked as log-only. See `sensor_module.py` line 291:
      ````python
        # This is just for logging! Do not use _ attributes for matching
        observed_state._semantic_id = semantic_id
      ```
Adding an `info` attribute to `State` is obviously trivial, as would be its encoder. Refactoring other code to use attributes in stored `State` objects rather than their corresponding lists in the buffer is a straightforward task, but it's a bit more involved.

Finally, note that I have not modified existing logging behavior for the goal-state-related items `goal_state_achieved` and `matching_step_when_output_goal_set`. Those values also live inside `GoalState.info` now, which makes them redundant. A future PR could eliminate this redundancy by retrieving those values from `GoalState` objects.

### CODE REVIEW

**Comment by scottcanoe** [comment]
Tagging @hlee9212 and @ramyamounir for visibility.

And one final note for @vkakerbeck and @nielsleadholm : The argument descriptions I added for `GoalState` are accurate in its current context (i.e., when a goal states is created by an `EvidenceGoalStateGenerator`), but I don't know if they are expected to hold more broadly. If not, I'm happy to address that now or in a future PR if/when that changes.

**Review by vkakerbeck — APPROVED** [review]
Looks nice :) Thanks for the change, this will be useful!

For historical context on why the buffer stores lists of the values of States instead of States: The `Buffer` class existed long before the `State` class. We actually added the `State` class rather late to the code and, for now, only added it at the interfaces between LMs and SMs. There is a follow-up item in our [project roadmap](https://docs.google.com/spreadsheets/d/1HlKqhMzYWxpW6_D0LtDd5qEuWRYHl6xMfxbIZ76lzng/edit?gid=0#gid=0) to "Use State class inside of LMs" (N20) which would include using it to store information in the buffer. It will be a larger refactor and we haven't gotten around to that yet as it hasn't been a top priority. @tristanls just tagging you to be aware of this item. It's nothing that would change functionality but it would increase readbility and clean up a lot of the code inside the LM and Buffer. I wouldn't prioritize it over our other planned refactors but if it starts causing issues with logging and generally modifying the code we should start thinking about it.

File: src/tbp/monty/frameworks/models/states.py [hunk 1]
```diff
@@ -217,45 +221,56 @@ class GoalState(State):
...
         self,
-        location,
-        morphological_features,
-        non_morphological_features,
-        confidence,
-        use_state,
-        sender_id,
-        sender_type,
-        goal_tolerances,
+        location: np.ndarray,
```
Comment (nielsleadholm) [inline src/tbp/monty/frameworks/models/states.py:237]: I believe we need something like the below to allow for `location` and `morphological_features` to be `None`:

```
        location: Optional[np.ndarray],
        morphological_features: Optional[Dict[str, Any]],
```

File: src/tbp/monty/frameworks/models/states.py [hunk 2]
```diff
@@ -217,45 +221,56 @@ class GoalState(State):
...
         Args:
-            location: ?
-            morphological_features: ?
-            non_morphological_features: ?
-            confidence: ?
-            use_state: ?
-            sender_id: ?
-            sender_type: ?
+            location: the location to move to in global/body-centric coordinates.
+              The location is near the target location on the object, but it has a
```
Comment (nielsleadholm) [inline src/tbp/monty/frameworks/models/states.py:251]: I would write this as, "For example, the location might be...", as this wouldn't necessarily be the approach for all model-based policies.

File: src/tbp/monty/frameworks/models/states.py [hunk 3]
```diff
@@ -217,45 +221,56 @@ class GoalState(State):
...
 
         Args:
-            location: ?
-            morphological_features: ?
-            non_morphological_features: ?
-            confidence: ?
-            use_state: ?
-            sender_id: ?
-            sender_type: ?
+            location: the location to move to in global/body-centric coordinates.
```
Comment (nielsleadholm) [inline src/tbp/monty/frameworks/models/states.py:250]: "...coordinates, or None, if the location is not specified as part of the goal state."

File: src/tbp/monty/frameworks/models/states.py [hunk 4]
```diff
@@ -217,45 +221,56 @@ class GoalState(State):
...
+              The location is near the target location on the object, but it has a
+              small offset that should place the agent/sensor at short distance from
+              the target location so that the sensor ends up at a reasonable viewing
+              distance from the target rather than on top of it.
+            morphological_features: dictionary of morphological features. Should include
```
Comment (nielsleadholm) [inline src/tbp/monty/frameworks/models/states.py:255]: "... or None."

**Review by nielsleadholm — COMMENTED** [review]
Beautiful, really like this as a way of including meta-data for logging while keeping it distinct from functional data. Makes a lot of sense if we are passing everything around as a state, that we can then append this information as required.

Just a couple comments on default values.

**Comment by scottcanoe** [comment]
@vkakerbeck Thanks, that makes a lot of sense. It'll be a nice refactor one of these days, and I imagine it'd help make Monty more generalized for other kinds of data as well.

@nielsleadholm Thanks for clarifying the documentation points. I've applied your edits.

**Review by nielsleadholm — APPROVED** [review]
LGTM!

**Review by tristanls — COMMENTED** [review]
Thank you for all the clear additions. I think there might be one issue with `goal_states` assignment, and I think the test should test through the serialization boundary.

**note**: My comment labels are intended to be interpreted per https://conventionalcomments.org/

File: src/tbp/monty/frameworks/models/goal_state_generation.py [hunk 5]
```diff
@@ -432,10 +433,15 @@ def _update_gsg_logging(self, output_goal_achieved: bool):
...
-            match_step = self.parent_lm.buffer.get_num_matching_steps() - 1
 
+            match_step = self.parent_lm.buffer.get_num_matching_steps() - 1
+            self.output_goal_state.info["achieved"] = output_goal_achieved
+            self.output_goal_state.info["matching_step_when_output_goal_set"] = (
+                match_step
+            )
             self.parent_lm.buffer.update_stats(
                 dict(
+                    goal_states=self.output_goal_state,
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply src/tbp/monty/frameworks/models/goal_state_generation.py:444]: Yes, it is intended since the keyword arguments control whether a list is created vs appended to. 

Line 72:
```python
        self.parent_lm.buffer.update_stats(
            dict(
                goal_states=[],
                matching_step_when_output_goal_set=[],
                goal_state_achieved=[],
            ),
            update_time=False,
            append=False,
            init_list=False,
        )
```
Line 442:
```python
            self.parent_lm.buffer.update_stats(
                dict(
                    goal_states=self.output_goal_state,
                    matching_step_when_output_goal_set=match_step,
                    goal_state_achieved=output_goal_achieved,
                ),
                update_time=False,
                append=True,
                init_list=True,
            )
```
In the first block, `append=False` means the buffer will set `self.stats["goal_states"]` to the given list, whereas `append=True` in the second block means it will append to that list. The code in this PR produces one list of goal states per episode which is the desired outcome.

**note:** ‏The parameters to `FeatureAtLocationBuffer.update_stats` are somewhat non-obvious though since the arguments co-mingle with each other and the state of the buffer's `stats` dictionary to decide what to do with the given data.

File: src/tbp/monty/frameworks/models/states.py [hunk 6]
```diff
@@ -311,3 +325,30 @@ def _check_all_attributes(self):
...
+        "confidence": goal_state.confidence,
+        "use_state": goal_state.use_state,
+        "sender_id": goal_state.sender_id,
+        "sender_type": goal_state.sender_type,
+        "goal_tolerances": goal_state.goal_tolerances,
+        "info": goal_state.info,
+    }
+
+
+BufferEncoder.register(GoalState, encode_goal_state)
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply src/tbp/monty/frameworks/models/states.py:354]: I'm not sure the best way either. Registering within `buffer.py` might cause circular imports in certain cases. My memory is fuzzy on this, but that did happen to me at one point, and the solution may have been to perform the registration in `states.py`.

File: tests/unit/frameworks/models/state_test.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                },
+                "achieved": False,
+                "matching_step_when_output_goal_set": None,
+            },
+        }
+        self.goal_state = GoalState(**self.goal_state_dict)
+
+    def test_encode(self):
+        encoded = encode_goal_state(self.goal_state)
+        self.assertDictEqual(encoded, self.goal_state_dict)
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply tests/unit/frameworks/models/state_test.py:56]: Fair point. I've added a test that compares the dumped -> loaded dictionary used to create the `GoalState` with the dumped -> loaded `GoalState` object. I think that resolves the issue.

File: src/tbp/monty/frameworks/models/goal_state_generation.py [hunk 8]
```diff
@@ -71,6 +71,7 @@ def reset(self):
         self._set_output_goal_state(self._generate_none_goal_state())
```
[the remark this answers is not in the document]
Comment (scottcanoe) [reply src/tbp/monty/frameworks/models/goal_state_generation.py:71]: I can't really speak to this much (maybe @nielsleadholm or @vkakerbeck have insight), but perhaps there was more planned for getter/setter methods, such as mutating the input before saving or returning it. Or they wanted to encourage their use in case those methods do have meaning in a future subclass. `output_goal_state` is public though, so I'm not sure.

**Review by tristanls — APPROVED** [review]
Thank you for the updates and explanations.
